### ISSUE TITLE
A nonstaff user can view all tickets?

### ISSUE THREAD

**Issue by andrewamanda** [django-helpdesk/django-helpdesk#326]
I set up the django_helpdesk. I assume by default a non-staff user can only view the tickets that he/she submitted.  But when I login as a non-staff user, and click on the Dashboard, I'm able to see all of the tickets created by other users and I can even update any ticket.

How do I set up so that I can only see the ticket submitted by myself?

Thank you
Andrew

**Comment by rossp** [comment]
It's assumed that Staff users are helpdesk staff, in that they can see/update/respond to any issue.

There's been prior discussion in ticket #239 about redefining the is_staff filter, but so far we don't have a solution and it should probably be reviewed again. I'd be happy to see it resolved :)

**Comment by andrewamanda** [comment]
Hi, Rossp,
I can understand that Staff users can see/update/respond to any issue. But my problem is that, a non-staff user login, he can still see under the dashboard all the tickets which were not created by him, thus he should not be able to see.

Is there any setting to disable a non-staff user from seeing tickets other than what was created by himself?

Thank you, Andrew

**Comment by maxmarsoft** [comment]
Hi Ross,

Have there been any updates on this issue? Last commented on in May this year. I am having the same problem in that a non staff user (user1) logs on and on their Dashboard they only see their tickets, but on the tickets page are listed tickets for all users not just those for user1.

I can see that the dashboard.html uses this code:
    {% if all_tickets_reported_by_current_user %}
    {% trans "All Tickets submitted by you" as ticket_list_caption %}
    {% include 'helpdesk/include/tickets.html' with ticket_list=all_tickets_reported_by_current_user ticket_list_empty_message="" %}
    {% endif %}
Which works, whereas the tickets page uses this:
    {% for ticket in tickets.object_list %}
which does not.

Any help to resolve this would be greatly appreciated.
Thanks, Max

**Comment by rossp** [comment]
Hi, Unfortunately I don't have time available right now to work on this so I am really going to need a pull request to make any changes - however can you please confirm your value for `settings.HELPDESK_ALLOW_NON_STAFF_TICKET_UPDATE`? If this is `True` then we should get the behaviour you're talking about.

**Comment by maxmarsoft** [comment]
Dear Ross,

Many thanks for taaking the time to reply. I can fully understand your time commitments and fully appreciate this.
For your info, the value for settings. HELPDESK_ALLOW_NON_STAFF_TICKET_UPDATE had been set to True as I thought this would allow customers to respond to tickets. I have now reset this to the default of False and it has not fixed the issue. User 2 can still see and respond to User1's tickets.

I will continue to try to find a fix for this but if you could spare a couple of minutes to give me some pointers that would be a great help.
Is there an online document that explains the logic behind this project that gives a more in-depth insight?
If not, can you possibly explain where the tickets.object_list in tickets_list.html and also all_tickets_reported_by_current_user in dashboard.html are generated to give me a chance of solving this issue.

Once again, many thanks for your time.

Kind Regards,

Max

Max Keogh
Systems Administrator

T : +44 (0) 1304 840506
F : +44 (0) 1304 840075
W : http://www.marinesoftware.co.uk

**Comment by maxmarsoft** [comment]
Hi Ross,

Apologies. Further to my last email, I hadn't restarted my web server when I tested the effect of changing the setting.
After restarting the web server I cannot view or add a ticket now as a non-staff user.

Max

**Comment by rossp** [comment]
Thanks for clarifying @maxmarsoft, glad you got it working.

**Comment by maxmarsoft** [comment]
Hi @Rossp. Sorry, I don't think I was clear enough in my last reply. It is the opposite of working. Now I cannot add or view tickets at all. Not even those for the logged in user. Now, only staff users can add or view tickets. I have even logged in to the admin page and given the non-staff users full privileges and still nothing.

**Comment by Nicoloren** [comment]
I think this issue is still open and I see the same behavior as Max: 

If HELPDESK_ALLOW_NON_STAFF_TICKET_UPDATE = True then a simple user can see a ticket created by another user with the url  _/helpdesk/tickets/%d/_

I didn't want to fork the project so my solution was top create a new view to list only ticket open by user with links.

Also, you'll need to override "ticket.html" to add some code to verify the user can see the ticket : 
`{% if ticket.submitter_email != request.user.email and not request.user.is_staff and not request.user.is_superuser %}`

This is not a very good solution but it seems to do the job...

**Comment by reduxionist** [comment]
@Nicoloren Thanks for the update on this issue, and for providing the work-around. We're currently transitioning to a new, expanded, project organization but I do agree that this is something we should address directly in a future release. Until that time, external PRs are, of course, always welcome. ;)

**Comment by maxkeogh** [comment]
Hi,
The way I got round this issue was to define whether the user should have access. In staff.py I defined the following function:

`def is_my_ticket(user, ticket):
    """Check to see if the user has permission to access
    a ticket. If not then deny access."""
    if user.is_superuser or user.is_staff or user.id == ticket.customer_id:
        return True
    else:
        return False
`

Then, in staff.py where ever access to a ticket is defined (eg, view_ticket, edit_ticket, delete_ticket, etc) I included these 2 lines of code:
`    if not is_my_ticket(request.user, ticket):
        raise PermissionDenied()
`
This has prevented a user access to any ticket where they are not authorized.

I hope this helps.

Max
